### ISSUE TITLE
TypeScript with TSX(JSX) behaves like normal ts file and reports syntax error

### ISSUE THREAD

**Issue by gaeulbyul** [kak-lsp/kak-lsp#211]
When I open `.tsx` file (valid syntax) with kakoune, kak-lsp shows syntax error(e.g. `'>' expected`, `Cannot find name 'div'`...).

Kakoune sets both  `.ts` and `.tsx` files to same filetype(`typescript`) unlike in Vim(`typescript`/`typescript.jsx`) or VSCode(`typescript`/`typescriptreact`). Maybe it is a cause of this issue?

* * *

- `kakoune`: `v2019.01.20-336-g125b1924` (build from git master branch)
- `kak-lsp`: `6.2.1` (build from git master branch)
- `typescript-language-server`: `0.3.8`
- OS: elementary OS 5.0 juno (based on Ubuntu 18.04)
- `kak-lsp.toml`:
```toml
[language.typescript]
filetypes = ["typescript"]
roots = ["tsconfig.json"]
command = "typescript-language-server"
args = ["--stdio"]
```

**Comment by ul** [comment]
Yes, this is the root cause of the problem. `typescript-language-server` treats tsx as a separate languageId. https://github.com/mawww/kakoune/blob/master/rc/filetype/javascript.kak needs to be updated to support `typescript.tsx` as a separate filetype. After that adding

```
[language."typescript.tsx"]
filetypes = ["typescript.tsx"]
roots = ["tsconfig.json"]
command = "typescript-language-server"
args = ["--stdio"]
```

to `kak-lsp.toml` should do the job (and if you are using `lsp-enable-window` then you need to list `typescript.tsx` in the hook pattern as well).

**Comment by brainshave** [comment]
I tried the suggested solution and it turns out that adding a separate entry for tsx files to `kak-lsp.toml` will make it start two instances of the TypeScript language server (one for `.ts` and one for `.tsx` files). This is less than ideal because if you have a `.ts` and `.tsx` file open, type changes are not propagated as quickly as if they were part of one instance. (Between instances, the only thing though which they can observe each other's changes is by observing the file system)

Before, the only time I had typescript server not treating `.tsx` as `.tsx` but a regular `.ts` was when the server is started and this file was the selected buffer. (Starting particular language servers is done lazily when a specific type is encountered.) So if I opened a regular `.ts` file first, waited for the TypeScript server to start, and then opened a `.tsx` file, it worked.

Playing with the suggestion above by splitting making Kakoune treat `.tsx` files as a separate type, I implemented it in https://github.com/brainshave/kakoune/commit/01270627ae9fe7346fe9631251a915cd0e0e2711 . The key to pick the right type on the server start turned out to be actually this part of the `kak-lsp.toml` file:

```toml
[language."typescript.tsx"]
```

I actually had it just `[language.tsx]` and it worked so it seems language server types in kak-lsp and kakoune file types are separate things. And it seems like the `tsx` string was passed to the TypeScript language server upon initialisation because whatever was after `language...` actually did make the difference for how the first file is treated. The actual kakoune file type seemed to not make a difference even after creating a special type for tsx:

```toml
filetypes = ["typescript-tsx"]
```

I haven't dig into specifics of "why", only relying here what worked for me :wink: 

So right now I'm testing a setup where I have only a **single** entry with:

```toml
[language.tsx]
filetypes = ["typescript"]
roots = ["package.json", "tsconfig.json"]
command = "npx"
args = ["typescript-language-server", "--stdio"]
```

This works out of the box with the non-patched versions of Kakoune v2019.07.01 and  kak-lsp 6.2.1. The only possible downside is that the first file opened is always going to be treated as a `.tsx` file even if it's `.ts` but I haven't found a situation where this caused a problem, yet :wink:
